style.md

```diff
@@ -0,0 +1,281 @@
+```julia
+# BAD!
+a < 0 && error("`a` parameter invalid")
+
+# GOOD!
+@assert a < 0 
+```
+- **Do not** use a `;` to call functions with keyword arguments unless it is required.  i.e. prefer `range(1, stop=5)` to `range(1; stop=5)`
+- **Use** the adjoint notation, `A'` instead of calling `transpose(A)` directly when working with real matrices (where they are identical).  With complex matrices, use whichever is appropriate.
+- **Use** the notation for stacking instead of the functions.  That is,
+```julia
+a = [1 2]
+b = [3 4]
+
+# BAD!
+hcat(a, b)
+vcat(a, b)
+
+# GOOD!
+[a b]
+[a; b] #or,
+[a;
+ b]
+```
+- **Use `I`**, the `UniformScaling` type, instead of constructing concrete matrices.
+```julia
+using LinearAlgebra
+A = [1 2; 3 4] 
+
+# BAD!
+Ival = [1 0; 0 1]
+A + Ival
+A + 2 * Ival
+
+# GOOD!
+A + I
+A + 2*I
+```
+- **Preallocate with `similar`** whenever possible, and avoid type annotations unless necessary.  The goal is to maintain code independent of types, which will aid later in generic programming.  Where you cannot, just allocate zeros, etc. instead 
+```julia
+N = 5
+x = [1.0, 2.0, 3.0, 4.0, 5.0]
+
+# Bad!
+y = Array{Float64}(undef, N)
+A = Array{Float64}(undef, N, N)
+
+# Better
+y = ones(N) #If we want the default, floats
+A = ones(N,N)
+
+# Best (if a candidate `x` exists)
+y = similar(x, N) #Keeps things generic.  The `N` here is not required if the same size
+A = similar(x, N, N) #Same type but NxN size
+```
+- **Don't use  `push!` when clearer alternatives exist** as it is harder for introductory reasoning and the size is preallocated.  But try to use broadcasting, comprehensions, etc. if clearer
+```julia
+# Bad!
+N = 5
+x = [] #Really bad since it is an Any vector!
+for i in 1:N
+    push!(x, 2.0 * i^2)
+end
+x
+
+# Better!
+x = ones(N)
+for i = 1:N
+    x[i] = 2.0 * i^2
+end
+x
+
+# Best!
+x = [2.0 * i^2 for i in 1:N]
+
+# or
+f(i) = 2.0 * i^2
+x = f.(1:5) #Use broadcasting
+```
+- Prefer `eachindex` to accessing the sizes of matrices.
+```julia
+x = [1.0, 2.0, 5.0, 2.1]
+y = similar(x)
+
+# BAD!
+for i in 1:length(x)
+    y[i] = x[i] + 2
+end
+y
+
+# BETTER!
+for i in eachindex(x)
+    y[i] = x[i] + 2
+end
+y
+
+# GOOD! #No way to preallocate y (although easier ways to write)
+x = rand(10)
+y = similar(x, 0) #Empty of same type as x
+for val in x
+    if val < 0.5
+        push!(y, val)
+    end
+end
+y
+```
+- **Use iterators directly** rather than accessing by index.
+```julia
+x = [1, 2, 3]
+
+# BAD!
+n = 0
+for i in eachindex(x)
+  n += x[i]^2
+end
+n
+
+# BETTER!
+n = 0
+for val in x
+  n += val^2
+end
+n
+
+# BEST!
+sum(xval -> xval^2, x) #i.e. transform each x and then reduce
+```
+
+- **Avoid `range` when possible** and use the `1.0:0.1:1.0` style notation, etc.  Getting rid of the `linspace` is a pain, but use ranges if the exact top doesn't matter.
+```julia
+Base.OneTo(5) # BAD!
```

Comment:
That was just me... probably not worth mentioning it. Might generate more confusion.